java-checks/src/main/java/org/sonar/java/checks/unused/UnusedLocalVariableCheck.java

```diff
@@ -19,128 +19,99 @@
-      BlockTree blockTree = (BlockTree) tree;
-      addVariables(blockTree.body());
-    } else if (tree.is(Tree.Kind.FOR_STATEMENT)) {
-      ForStatementTree forStatementTree = (ForStatementTree) tree;
-      addVariables(forStatementTree.initializer());
-    } else if (tree.is(Tree.Kind.FOR_EACH_STATEMENT)) {
-      ForEachStatement forEachStatement = (ForEachStatement) tree;
-      addVariables(Collections.singletonList(forEachStatement.variable()));
-    } else if (tree.is(Tree.Kind.EXPRESSION_STATEMENT)) {
-      leaveExpressionStatement((ExpressionStatementTree) tree);
-    } else {
-      checkVariableUsages();
-      variables.clear();
-      assignments.clear();
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.COMPILATION_UNIT)) {
+      UNRESOLVED_IDENTIFIERS_VISITOR.check(tree);
+      unusedVariables.clear();
     }
   }
 
-  private void leaveExpressionStatement(ExpressionStatementTree expressionStatement) {
-    ExpressionTree expression = expressionStatement.expression();
-    if (expression.is(ASSIGNMENT_KINDS)) {
-      addAssignment(((AssignmentExpressionTree) expression).variable());
-    } else if (expression.is(INCREMENT_KINDS)) {
-      addAssignment(((UnaryExpressionTree) expression).expression());
+  @Override
+  public void leaveNode(Tree tree) {
+    switch (tree.kind()) {
+      case IDENTIFIER:
+        IdentifierTree ident = (IdentifierTree) tree;
+        if (isRValue(ident)) {
+          unusedVariables.remove(ident.symbol());
+        }
+        break;
+      case VARIABLE:
+        addVariable((VariableTree) tree);
+        break;
+      case PATTERN_INSTANCE_OF:
+        addVariable(((PatternInstanceOfTree) tree).variable());
+        break;
+      default: // COMPILATION_UNIT
+        for (Map.Entry<Symbol, VariableTree> entry : unusedVariables.entrySet()) {
+          VariableTree variable = entry.getValue();
+          reportIssue(variable.simpleName(), "Remove this unused \"" + variable.symbol().name() + "\" local variable.");
+        }
+        break;
     }
   }
 
-  private void checkVariableUsages() {
-    for (VariableTree variableTree : variables) {
-      Symbol symbol = variableTree.symbol();
-      if (symbol.usages().size() == assignments.getOrDefault(symbol, Collections.emptyList()).size()) {
-        reportIssue(variableTree.simpleName(), "Remove this unused \"" + symbol.name() + "\" local variable.");
-      }
+  /**
+   * An identifier is being used as an r-value if it is not used as the left operand of an assignment nor as the
+   * operand of a stand-alone increment
+   */
+  private static boolean isRValue(IdentifierTree tree) {
+    if (tree.parent() instanceof AssignmentExpressionTree) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) tree.parent();
+      return assignment.variable() != tree;
     }
+    return !(tree.parent().is(INCREMENT_KINDS) && tree.parent().parent().is(Tree.Kind.EXPRESSION_STATEMENT));
```

Comment:
what about parenthesized expressions in parent? we can probably s kip them.

Reply:
Done